File: web/apps/app/lib/core/chat/__tests__/subscription-capability-reachability.test.ts
// @vitest-environment jsdom
//
// Capability reachability for the subscription path (Codex / Grok subscriptions).
//
// Subscription models are not in the metadata catalog and never carry v2
// `capabilityControls`, so the web toggle and the reasoning tier pills render nowhere and
// nothing is ever injected outbound. An outbound implementation that copies the upstream
// declaration is dead code as long as that gate stays shut upstream.
//
// These cases pin the full chain once that gate is opened:
//   upstream /models declaration -> control verdict -> preference identity -> outbound intent
// Any link falling back to fail-closed turns this file red.

import { describe, expect, it, vi } from 'vitest';
import type { AIModel, Provider } from '@oriveo/shared';

vi.mock('../../metadata/metadata-client', () => ({
  // The core fact about subscription models: they are not in the catalog.
  resolveCatalogModel: () => null,
  getModelTransport: () => undefined,
  getCapabilityRuntime: () => undefined,
  getMetadataRevision: () => undefined,
  getRelayRuntimeConfig: () => undefined,
  resolveGenerationProfileRef: () => undefined,
  getDeclaredReasoningLevels: () => [],
  getDeclaredReasoningDefaultLevel: () => undefined,
  subscriptionDeclaredReasoningLevels: (_providerKind: string, model: AIModel) =>
    model.upstreamReasoningLevels ?? [],
}));

vi.mock('../../../infra/storage/partition', () => ({ getActiveUIDSync: () => 'uid-1' }));

import { presentCapabilityControl } from '../capability-control-presentation';
import { webPreferenceReachesTheWire } from '../capability-control-presentation';
import { capabilityRuntimeIdentity } from '../capability-preference-settings';
import { filterRequestCapabilityIntent } from '../stream-options';

function codexProvider(): Provider {
  return {
    id: '11111111-1111-4111-8111-111111111111', kind: 'openAI', status: { kind: 'connected' },
    models: [], catalogModels: [], apiKey: '', apiKeyPreview: '',
    authMode: 'subscription',
    openAISubscription: { accessToken: 'at', accountID: 'acc-1', obtainedAt: 1 },
  } as unknown as Provider;
}

function grokProvider(): Provider {
  return {
    id: '22222222-2222-4222-8222-222222222222', kind: 'grok', status: { kind: 'connected' },
    models: [], catalogModels: [], apiKey: '', apiKeyPreview: '',
    authMode: 'subscription',
    grokSubscription: { accessToken: 'at', obtainedAt: 1 },
  } as unknown as Provider;
}

/** The shape built by `buildOpenAISubscriptionModels`: capability bits and the tier table are copied from the upstream /models declaration. */
function subscriptionModel(overrides: Partial<AIModel> = {}): AIModel {
  return {
    id: 'gpt-5.6-sol', name: 'GPT-5.6-Sol',
    capabilities: ['text', 'web', 'reasoning'],
    upstreamReasoningLevels: ['low', 'medium', 'high'],
    reasoningModeAvailable: true, isAvailable: true, isDefault: true, priceTier: '',
    ...overrides,
  } as unknown as AIModel;
}

describe('subscription path: control verdicts follow the upstream declaration', () => {
  it('makes the web control usable when upstream declares web search, rather than showing it as unadjustable', () => {
    // Falling back to unknown would render "upstream says it is supported" as a line of status text with no toggle to reach.
    const control = presentCapabilityControl(codexProvider(), subscriptionModel(), 'web');
    expect(control.state).toBe('auto_available');
    expect(control.availableIntents).toContain('automatic');
  });

  it('reports unavailable (definitely unsupported) rather than unknown when upstream declares no web search', () => {
    const control = presentCapabilityControl(
      codexProvider(),
      subscriptionModel({ capabilities: ['text'] }),
      'web',
    );
    expect(control.state).toBe('unavailable');
    expect(control.reasonCode).toBe('model_capability_absent');
  });

  it('computes available tiers with the same function used outbound: only tiers that map to a declared upstream value appear', () => {
    const control = presentCapabilityControl(codexProvider(), subscriptionModel(), 'reasoning');
    expect(control.state).toBe('auto_available');
    // Upstream declares only low/medium/high: max degrades to high, and the contract
    // guarantees degradation still lands on a legal value rather than silently sending
    // nothing, so it is a tier that really can go out.
    expect(control.availableIntents).toEqual(['low', 'balanced', 'deep', 'max']);
  });

  it('hides tiers that cannot be mapped when upstream declares a single one, rather than offering a control that does nothing', () => {
    // With only xhigh declared, fast maps to low/minimal/medium and matches none of them, so the tier must not appear.
    const control = presentCapabilityControl(
      codexProvider(),
      subscriptionModel({ upstreamReasoningLevels: ['xhigh'] }),
      'reasoning',
    );
    expect(control.availableIntents).toEqual(['max']);
  });

  it('reports the reasoning tier as unavailable when upstream declares no tier table, rather than offering a control that does nothing', () => {
    const control = presentCapabilityControl(
      codexProvider(),
      subscriptionModel({ upstreamReasoningLevels: [] }),
      'reasoning',
    );
    expect(control.state).toBe('unavailable');
    expect(control.reasonCode).toBe('upstream_parameter_not_declared');
  });

  it('keeps Grok subscription Responses web search constantly automatic', () => {
    const control = presentCapabilityControl(
      grokProvider(), subscriptionModel({ upstreamApiBackend: 'responses' }), 'web',
    );
    expect(control.state).toBe('auto_available');
    expect(control.availableIntents).toEqual(['automatic']);
  });

  it('keeps reasoning tiers usable for a Grok subscription, since reasoning_effort is known to be accepted', () => {
    const control = presentCapabilityControl(grokProvider(), subscriptionModel(), 'reasoning');
    expect(control.state).toBe('auto_available');
    expect(control.availableIntents.length).toBeGreaterThan(0);
  });
});

describe('subscription path: preference identity is available so choices can be stored', () => {
  it('still resolves an identity for a subscription model that has no catalog transport', () => {
    // With a null identity the capability panel is entirely read-only and the user's choice can neither be stored nor read back.
    const identity = capabilityRuntimeIdentity(codexProvider(), subscriptionModel());
    expect(identity).not.toBeNull();
    expect(identity?.canonicalModelId).toBe('gpt-5.6-sol');
    // Codex uses /responses and Grok uses /chat/completions, so the transport for this path
    // is definite and needs no confirmation from the catalog.
    expect(identity?.finalTransport).toBe('openai_responses');
    expect(identity?.transportIdentity).toBeTruthy();
  });

  it('defaults a Grok subscription to Responses when nothing is declared', () => {
    const identity = capabilityRuntimeIdentity(grokProvider(), subscriptionModel());
    expect(identity?.finalTransport).toBe('openai_responses');
  });

  it('keeps the two paths on distinct identities so preferences do not leak across them', () => {
    const codex = capabilityRuntimeIdentity(codexProvider(), subscriptionModel());
    const grok = capabilityRuntimeIdentity(grokProvider(), subscriptionModel());
    expect(codex?.providerId).not.toBe(grok?.providerId);
  });
});

describe('subscription path: user intent really reaches the outbound request', () => {
  it('carries web intent through to the outbound gate', () => {
    localStorage.clear();
    const provider = codexProvider();
    const model = subscriptionModel();
    const identity = capabilityRuntimeIdentity(provider, model);
    // The second gate: whether the intent can reach the wire. It fails closed when the identity is null or the control is unavailable.
    const reaches = webPreferenceReachesTheWire({
      provider, model, transportIdentity: identity?.transportIdentity ?? '',
    });
    expect(reaches).toBe(true);

    const intent = filterRequestCapabilityIntent({
      provider, model, reasoningMode: 'automatic', webSearchEnabled: true,
      webReachesTheWire: reaches,
    });
    // A false here is the bug where the user turns web search on and nothing is ever sent.
    expect(intent.supportsWebSearch).toBe(true);
  });

  it('does not let web intent reach the outbound request for a model upstream never declared it for', () => {
    localStorage.clear();
    const provider = codexProvider();
    const model = subscriptionModel({ capabilities: ['text'] });
    const identity = capabilityRuntimeIdentity(provider, model);
    const reaches = webPreferenceReachesTheWire({
      provider, model, transportIdentity: identity?.transportIdentity ?? '',
    });
    expect(reaches).toBe(false);

    const intent = filterRequestCapabilityIntent({
      provider, model, reasoningMode: 'automatic', webSearchEnabled: true,
      webReachesTheWire: reaches,
    });
    expect(intent.supportsWebSearch).toBe(false);
  });

  it('carries the reasoning tier the user picked through to the outbound intent', () => {
    localStorage.clear();
    const provider = codexProvider();
    const model = subscriptionModel();
    const intent = filterRequestCapabilityIntent({
      provider, model, reasoningMode: 'deep', webSearchEnabled: false,
    });
    expect(intent.reasoning).toBe('deep');
  });

  it('injects no tier for automatic, leaving it to the upstream default_reasoning_level', () => {
    localStorage.clear();
    const provider = codexProvider();
    const model = subscriptionModel();
    const intent = filterRequestCapabilityIntent({
      provider, model, reasoningMode: 'automatic', webSearchEnabled: false,
    });
    expect(intent.reasoning).toBeUndefined();
  });
});

describe('subscription path: API key mode is unaffected', () => {
  it('keeps the same OpenAI provider on the catalog checks in API key mode, so the subscription branch does not bleed over', () => {
    const provider = { ...codexProvider(), authMode: 'apiKey' } as Provider;
    // Not found in the catalog (the mock returns null) means an honest unknown, exactly as before subscriptions existed.
    const control = presentCapabilityControl(provider, subscriptionModel(), 'web');
    expect(control.state).toBe('unknown');
  });
});

File: android/app/src/test/java/ai/oriveo/community/core/provider/SubscriptionCapabilityReachabilityTest.kt
package ai.oriveo.community.core.provider

import ai.oriveo.community.core.data.remote.canonicalCapabilityTransport
import ai.oriveo.community.core.model.AIModel
import ai.oriveo.community.core.model.ModelCapability
import ai.oriveo.community.core.model.Provider
import ai.oriveo.community.core.model.ProviderAuthMode
import ai.oriveo.community.core.model.ProviderKind
import ai.oriveo.community.core.provider.transport.TransportKind
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Capability reachability on the subscription paths (Codex and Grok subscription sign-in).
 *
 * These came directly out of an investigation: subscription models are not in the model catalog, so
 * they never carry a `capabilityControls` block, and as a result the web search switch and the
 * reasoning level picker rendered nowhere and nothing was ever injected outbound. The carefully
 * written "copy what upstream declared" implementation had become dead code, because the fix lived
 * downstream of a gate that closed upstream of it.
 *
 * What this suite pins is the full chain once that gate is opened:
 *   upstream /models declaration -> capability verdict -> preference identity
 * If any link falls back to fail-closed behaviour, these turn red.
 */
class SubscriptionCapabilityReachabilityTest {

    /** A subscription connection: the test is `authMode`, not providerKind, because the same OpenAI provider has a different source of truth for capabilities in each mode. */
    private fun subscriptionProvider(kind: ProviderKind = ProviderKind.OpenAI) = Provider(
        id = if (kind == ProviderKind.Grok) "22222222-2222-4222-8222-222222222222"
            else "11111111-1111-4111-8111-111111111111",
        kind = kind,
        authMode = ProviderAuthMode.Subscription,
    )

    /** The shape a subscription catalog entry is modelled into: both the capability bits and the level table are copied from what upstream declares at `/models`. */
    private fun subscriptionModel(
        id: String = "gpt-5.6-sol",
        capabilities: List<ModelCapability> = listOf(
            ModelCapability.Text, ModelCapability.Web, ModelCapability.Reasoning,
        ),
        declaredLevels: List<String> = listOf("low", "medium", "high"),
        apiBackend: String? = null,
    ) = AIModel(
        id = id,
        name = id,
        capabilities = capabilities,
        reasoningModeAvailable = capabilities.contains(ModelCapability.Reasoning),
        upstreamReasoningLevels = declaredLevels,
        upstreamApiBackend = apiBackend,
    )

    // ── The capability verdict follows what upstream declares ──

    @Test
    fun `a declared web search capability leaves the control usable rather than degraded to not adjustable`() {
        val verdict = CapabilityControlResolution.resolve(
            provider = subscriptionProvider(), model = subscriptionModel(), capability = "web",
        )
        assertEquals("auto_available", verdict.state)
        assertTrue(verdict.intents.contains("automatic"))
        assertEquals("subscription_upstream_declared", verdict.reasonCode)
    }

    @Test
    fun `an undeclared web search capability is unavailable, a definite no, not unknown meaning not looked into yet`() {
        val verdict = CapabilityControlResolution.resolve(
            provider = subscriptionProvider(),
            model = subscriptionModel(capabilities = listOf(ModelCapability.Text)),
            capability = "web",
        )
        assertEquals("unavailable", verdict.state)
        assertEquals("model_capability_absent", verdict.reasonCode)
    }

    @Test
    fun `the offered levels come from the same function the outbound path uses, so only levels that map onto a declared value appear`() {
        val verdict = CapabilityControlResolution.resolve(
            provider = subscriptionProvider(), model = subscriptionModel(), capability = "reasoning",
        )
        assertEquals("auto_available", verdict.state)
        // Upstream declares only low/medium/high, so max walks the fallback chain down to high. The
        // fallback is guaranteed to land on a legal value rather than silently sending nothing, so
        // max is still a level that really can go out.
        assertEquals(listOf("low", "balanced", "deep", "max"), verdict.intents)
    }

    @Test
    fun `with only one declared level, the levels that map onto nothing do not appear`() {
        // Only xhigh is declared, and the fast candidates are low/minimal/medium, none of which
        // match, so that level must not be offered.
        val verdict = CapabilityControlResolution.resolve(
            provider = subscriptionProvider(),
            model = subscriptionModel(declaredLevels = listOf("xhigh")),
            capability = "reasoning",
        )
        assertEquals(listOf("max"), verdict.intents)
    }

    @Test
    fun `with no declared level table the reasoning control is unavailable rather than a button that does nothing`() {
        val verdict = CapabilityControlResolution.resolve(
            provider = subscriptionProvider(),
            model = subscriptionModel(declaredLevels = emptyList()),
            capability = "reasoning",
        )
        assertEquals("unavailable", verdict.state)
        assertEquals("upstream_parameter_not_declared", verdict.reasonCode)
    }

    @Test
    fun `a Grok subscription on Responses that declares web search offers automatic only`() {
        val verdict = CapabilityControlResolution.resolve(
            provider = subscriptionProvider(kind = ProviderKind.Grok),
            model = subscriptionModel(id = "grok-4.6", apiBackend = "responses"),
            capability = "web",
        )
        assertEquals("auto_available", verdict.state)
        assertEquals(listOf("automatic"), verdict.intents)
    }

    @Test
    fun `reasoning levels on a Grok subscription are usable as normal`() {
        val verdict = CapabilityControlResolution.resolve(
            provider = subscriptionProvider(kind = ProviderKind.Grok),
            model = subscriptionModel(id = "grok-4.6"),
            capability = "reasoning",
        )
        assertEquals("auto_available", verdict.state)
        assertTrue(verdict.intents.isNotEmpty())
    }

    @Test
    fun `every other capability on a subscription path is honestly unknown rather than pretending to be unsupported`() {
        val verdict = CapabilityControlResolution.resolve(
            provider = subscriptionProvider(), model = subscriptionModel(), capability = "generation",
        )
        assertEquals("unknown", verdict.state)
    }

    // ── Panel presentation, the second gate: auto_available must not be knocked back to unknown by the transport check ──

    @Test
    fun `auto_available on a subscription path survives the exact-transport check and reaches the panel`() {
        // Subscription models have no recipe, so `hasExactTransportRecipe` is necessarily false.
        // Without an exemption, `status(...)` would degrade every auto_available back to Unknown and
        // opening the verdict gate would have achieved nothing.
        val presentation = CapabilityControlPresentationResolver.presentation(
            provider = subscriptionProvider(), model = subscriptionModel(), capability = "web",
        )
        assertEquals(CapabilityControlPresentation.AutomaticAvailable, presentation)
        // The third gate has the same source: the globe lit up in the panel has to be the one that
        // really goes out on the wire.
        assertTrue(CapabilityWebPreferenceLiveness.reachesTheWire(presentation, customIsActive = false))
    }

    @Test
    fun `a capability upstream did not declare shows as unsupported in the panel rather than unknown`() {
        val presentation = CapabilityControlPresentationResolver.presentation(
            provider = subscriptionProvider(),
            model = subscriptionModel(capabilities = listOf(ModelCapability.Text)),
            capability = "web",
        )
        assertEquals(CapabilityControlPresentation.Unsupported, presentation)
    }

    // ── Preference identity, so a choice can actually be stored ──

    @Test
    fun `a subscription model resolves no catalog transport yet still has a valid identity`() {
        // With a null identity the whole capability panel goes read-only and not a single byte of the
        // user's level choice or web search switch can be stored.
        val identity = ModelControlRuntimeIdentityResolver.resolve(
            provider = subscriptionProvider(), model = subscriptionModel(),
        )
        assertNotNull(identity)
        assertEquals("gpt-5.6-sol", identity?.canonicalModelId)
        assertEquals(
            canonicalCapabilityTransport(TransportKind.OpenAIResponses.wireValue),
            identity?.finalTransport,
        )
        assertEquals(
            ModelControlRuntimeIdentityResolver.SUBSCRIPTION_RUNTIME_REVISION,
            identity?.runtimeRevision,
        )
    }

    @Test
    fun `a Grok subscription with no declaration defaults to Responses, avoiding a fake search`() {
        val identity = ModelControlRuntimeIdentityResolver.resolve(
            provider = subscriptionProvider(kind = ProviderKind.Grok),
            model = subscriptionModel(id = "grok-4.6"),
        )
        assertEquals(
            canonicalCapabilityTransport(TransportKind.OpenAIResponses.wireValue),
            identity?.finalTransport,
        )
    }

    @Test
    fun `an explicit chat declaration on a Grok subscription must be honoured and web search is then unavailable`() {
        val provider = subscriptionProvider(kind = ProviderKind.Grok)
        val model = subscriptionModel(id = "grok-legacy", apiBackend = "chat")
        assertEquals(TransportKind.OpenAIChat.wireValue,
            CapabilityControlResolution.subscriptionFinalTransport(provider, model))
        assertEquals("unavailable",
            CapabilityControlResolution.resolve(provider, model, "web").state)
    }

    @Test
    fun `the two paths have different identities so a preference cannot bleed from one to the other`() {
        val codex = ModelControlRuntimeIdentityResolver.resolve(
            provider = subscriptionProvider(), model = subscriptionModel(),
        )
        val grok = ModelControlRuntimeIdentityResolver.resolve(
            provider = subscriptionProvider(kind = ProviderKind.Grok), model = subscriptionModel(),
        )
        // storageIdentity is deliberately transport+revision only; provider/connection id is the
        // store's separate partition column. Compare the complete runtime identities here.
        assertTrue(codex != grok)
    }

    // ── The API key mode is not caught in the crossfire ──

    @Test
    fun `the same OpenAI provider in API key mode does not enter the subscription branch`() {
        val provider = Provider(id = "byok", kind = ProviderKind.OpenAI, authMode = ProviderAuthMode.ApiKey)
        assertTrue(!CapabilityControlResolution.isSubscriptionLink(provider))
        assertNull(CapabilityControlResolution.subscriptionFinalTransport(provider))
    }
}
